File: electron/src/modules/game/camera.js
export class Camera {
    constructor(width, height) {
        this.x = 0;
        this.y = 0;
        this.width = width;
        this.height = height;
        this.zoom = 1;
        this.minZoom = 0.5;
        this.maxZoom = 3.0;
        this.zoomSpeed = 0.1;
        this.followSpeed = 0.1;
        this.targetZoom = 1;
        
        // Camera mode support
        this.mode = 'fixed-angle'; // 'fixed-angle' or 'player-perspective'
        this.rotation = 0; // Camera rotation for fixed-angle mode (radians)
        this.rotationSpeed = Math.PI * 0.8; // radians per second
    }

    // Set camera mode
    setMode(mode) {
        if (mode === 'fixed-angle' || mode === 'player-perspective') {
            this.mode = mode;
            console.log(`[Camera] Mode set to: ${mode}`);
        }
    }

    // Toggle between camera modes
    toggleMode() {
        this.mode = this.mode === 'fixed-angle' ? 'player-perspective' : 'fixed-angle';
        console.log(`[Camera] Mode toggled to: ${this.mode}`);
        return this.mode;
    }

    // Rotate camera in fixed-angle mode
    rotateCamera(deltaRotation) {
        if (this.mode === 'fixed-angle') {
            this.rotation += deltaRotation;
            // Keep rotation in reasonable bounds
            while (this.rotation > Math.PI * 2) this.rotation -= Math.PI * 2;
            while (this.rotation < 0) this.rotation += Math.PI * 2;
        }
    }

    // Reset camera rotation to north (0 degrees)
    resetRotation() {
        this.rotation = 0;
        console.log('[Camera] Rotation reset to north');
    }

    // Handle mouse wheel zoom
    handleZoom(delta) {
        // Calculate zoom change based on wheel delta
        // delta > 0 means scroll down (zoom out), delta < 0 means scroll up (zoom in)
        const zoomChange = delta > 0 ? 0.9 : 1.1;
        this.targetZoom = Math.max(this.minZoom, Math.min(this.maxZoom, this.zoom * zoomChange));
    }

    // Update zoom with smooth transitions
    update(deltaTime) {
        // Smooth zoom transition
        if (Math.abs(this.zoom - this.targetZoom) > 0.01) {
            this.zoom += (this.targetZoom - this.zoom) * this.zoomSpeed;
        }
    }

    // Follow a target with smooth movement
    follow(targetX, targetY) {
        this.x += (targetX - this.x) * this.followSpeed;
        this.y += (targetY - this.y) * this.followSpeed;
    }

    // Apply camera transform to context
    applyTransform(ctx) {
        ctx.save();
        
        // Apply zoom and translation
        ctx.translate(this.width / 2, this.height / 2);
        ctx.scale(this.zoom, this.zoom);
        
        if (this.mode === 'player-perspective') {
            // In player-perspective mode, we'll apply rotation in the game render loop
            // based on player angle, so we just translate here
            ctx.translate(-this.x, -this.y);
        } else {
            // Fixed-angle mode: apply camera rotation
            ctx.rotate(-this.rotation);
            ctx.translate(-this.x, -this.y);
        }
    }

    // Apply player-perspective transform (called from game render loop)
    applyPlayerPerspectiveTransform(ctx, playerAngle) {
        if (this.mode === 'player-perspective') {
            // Rotate the entire world around the player's position
            // First translate to player position, rotate, then translate back
            ctx.translate(this.x, this.y);
            ctx.rotate(-playerAngle);
            ctx.translate(-this.x, -this.y);
        }
    }

    // Restore camera transform
    restoreTransform(ctx) {
        ctx.restore();
    }

    // Convert screen coordinates to world coordinates
    screenToWorld(screenX, screenY, playerAngle = 0) {
        let worldX, worldY;
        
        // Undo zoom and center translation
        const x = (screenX - this.width / 2) / this.zoom;
        const y = (screenY - this.height / 2) / this.zoom;
        const undoAngle = this.mode == 'player-perspective' ? playerAngle : this.rotation;
        
        // Undo rotation
        const cos = Math.cos(undoAngle);
        const sin = Math.sin(undoAngle);
        worldX = x * cos - y * sin + this.x;
        worldY = x * sin + y * cos + this.y;
        
        return { x: worldX, y: worldY };
    }

    // Convert world coordinates to screen coordinates
    worldToScreen(worldX, worldY, playerAngle = 0) {
        let screenX, screenY;
        
        // Translate to camera-relative coordinates
        const x = worldX - this.x;
        const y = worldY - this.y;
        const angle = this.mode === 'player-perspective' ? playerAngle : this.rotation; 

        const cos = Math.cos(-angle);
            const sin = Math.sin(-angle);
            screenX = (x * cos - y * sin) * this.zoom + this.width / 2;
            screenY = (x * sin + y * cos) * this.zoom + this.height / 2;
        
        return { x: screenX, y: screenY };
    }

    // Get camera bounds in world coordinates
    getBounds() {
        const halfWidth = this.width / (2 * this.zoom);
        const halfHeight = this.height / (2 * this.zoom);
        
        return {
            left: this.x - halfWidth,
            right: this.x + halfWidth,
            top: this.y - halfHeight,
            bottom: this.y + halfHeight,
            width: this.width / this.zoom,
            height: this.height / this.zoom
        };
    }

    // Resize camera for new window size
    resize(width, height) {
        this.width = width;
        this.height = height;
        console.log('[Camera] Camera resized:', { width, height });
    }

    // Get camera info for debugging
    getInfo() {
        return {
            position: `(${this.x.toFixed(1)}, ${this.y.toFixed(1)})`,
            zoom: `${this.zoom.toFixed(2)}x`,
            mode: this.mode,
            rotation: `${(this.rotation * 180 / Math.PI).toFixed(1)}°`,
            bounds: this.getBounds()
        };
    }
} 

File: javafx/src/main/java/com/game/rendering/Camera.java
package com.game.rendering;

import javafx.scene.canvas.GraphicsContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Camera {
    private static final Logger logger = LoggerFactory.getLogger(Camera.class);
    
    private double x, y;
    private double zoom = 1.0;
    private double targetZoom = 1.0;
    private double rotation = 0.0;
    private double width, height;
    private CameraMode mode = CameraMode.FIXED_ANGLE;
    
    public enum CameraMode {
        FIXED_ANGLE,
        PLAYER_PERSPECTIVE
    }
    
    public Camera(double width, double height) {
        this.width = width;
        this.height = height;
        logger.info("Camera initialized: {}x{}", width, height);
    }
    
    public void update(double deltaTime) {
        // Smooth zoom interpolation
        double zoomDiff = targetZoom - zoom;
        if (Math.abs(zoomDiff) > 0.01) {
            zoom += zoomDiff * deltaTime * 5.0; // Smooth interpolation
        }
        
        // Clamp zoom
        zoom = Math.max(0.1, Math.min(5.0, zoom));
    }
    
    public void applyTransform(GraphicsContext gc) {
        gc.save();
        
        // Apply camera transformations
        gc.translate(width / 2, height / 2);
        gc.scale(zoom, zoom);
        gc.rotate(Math.toDegrees(rotation));
        gc.translate(-x, -y);
    }
    
    public void restoreTransform(GraphicsContext gc) {
        gc.restore();
    }
    
    public void follow(double targetX, double targetY) {
        this.x = targetX;
        this.y = targetY;
    }
    
    public void setZoom(double zoom) {
        this.targetZoom = Math.max(0.1, Math.min(5.0, zoom));
    }
    
    public void setRotation(double rotation) {
        this.rotation = rotation;
    }
    
    public void setMode(CameraMode mode) {
        this.mode = mode;
        logger.info("Camera mode changed to: {}", mode);
    }
    
    public void resize(double width, double height) {
        this.width = width;
        this.height = height;
    }
    
    // Getters
    public double getX() { return x; }
    public double getY() { return y; }
    public double getZoom() { return zoom; }
    public double getRotation() { return rotation; }
    public CameraMode getMode() { return mode; }
    public double getWidth() { return width; }
    public double getHeight() { return height; }
} 